Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscSignatureValidator.kt

```diff
@@ -2,16 +2,144 @@ package de.rki.coronawarnapp.covidcertificate.signature.core
 
 import dagger.Reusable
 import de.rki.coronawarnapp.covidcertificate.common.certificate.DccData
+import de.rki.coronawarnapp.covidcertificate.common.certificate.DscMessage
+import de.rki.coronawarnapp.covidcertificate.common.certificate.DscMessage.Algorithm.ES256
+import de.rki.coronawarnapp.covidcertificate.common.certificate.DscMessage.Algorithm.PS256
+import de.rki.coronawarnapp.covidcertificate.common.certificate.RecoveryDccV1
+import de.rki.coronawarnapp.covidcertificate.common.certificate.TestDccV1
+import de.rki.coronawarnapp.covidcertificate.common.certificate.VaccinationDccV1
+import de.rki.coronawarnapp.covidcertificate.common.exception.InvalidHealthCertificateException
+import de.rki.coronawarnapp.covidcertificate.common.exception.InvalidHealthCertificateException.ErrorCode.HC_DSC_EXPIRED
+import de.rki.coronawarnapp.covidcertificate.common.exception.InvalidHealthCertificateException.ErrorCode.HC_DSC_NOT_YET_VALID
+import de.rki.coronawarnapp.covidcertificate.common.exception.InvalidHealthCertificateException.ErrorCode.HC_DSC_NO_MATCH
+import de.rki.coronawarnapp.covidcertificate.common.exception.InvalidHealthCertificateException.ErrorCode.HC_DSC_OID_MISMATCH_RC
+import de.rki.coronawarnapp.covidcertificate.common.exception.InvalidHealthCertificateException.ErrorCode.HC_DSC_OID_MISMATCH_TC
+import de.rki.coronawarnapp.covidcertificate.common.exception.InvalidHealthCertificateException.ErrorCode.HC_DSC_OID_MISMATCH_VC
+import kotlinx.coroutines.flow.first
+import org.bouncycastle.jce.provider.BouncyCastleProvider
 import timber.log.Timber
+import java.security.PublicKey
+import java.security.Security
+import java.security.Signature
+import java.security.cert.CertificateExpiredException
+import java.security.cert.CertificateFactory
+import java.security.cert.CertificateNotYetValidException
+import java.security.cert.X509Certificate
 import javax.inject.Inject
 
 @Reusable
-class DscSignatureValidator @Inject constructor() {
+class DscSignatureValidator @Inject constructor(
+    private val dscRepository: DscRepository
+) {
 
-    suspend fun isSignatureValid(dscData: DscData, certificateData: DccData<*>): Boolean {
-        Timber.tag(TAG).d("isSignatureValid(certificateData=%s)", certificateData.header)
-        // TODO
-        return true
+    init {
+        Security.addProvider(BouncyCastleProvider()) // For SHA256withRSA/PSS
+    }
+
+    private val vcOids = setOf(
+        "1.3.6.1.4.1.1847.2021.1.2",
+        "1.3.6.1.4.1.0.1847.2021.1.2"
+    )
+    private val tcOids = setOf(
+        "1.3.6.1.4.1.1847.2021.1.1",
+        "1.3.6.1.4.1.0.1847.2021.1.1"
+    )
+    private val rcOids = setOf(
+        "1.3.6.1.4.1.1847.2021.1.3",
+        "1.3.6.1.4.1.0.1847.2021.1.3"
+    )
+
+    private val oidSet = vcOids + tcOids + rcOids
+
+    /**
+     * @throws InvalidHealthCertificateException if validation fail, otherwise it is OK!
+     */
+    suspend fun validateSignature(dccData: DccData<*>) {
```

Comment:
I would like it if we could rename this `requireValidSignature` like the other apicalls that either do nothing or throw an exception, e.g. `requireNotNull(....)`.

What do you think?

Reply:
I don't think `require*` is a suitable name

Reply:
Pretty common though? `requireContext/require()/requireNotNull`

Not a blocker for me though.